### PR TITLE
Add support for Dual and Neutral classes on Duels stats

### PR DESCRIPTION
**Pull request by edipo2s**
- [ ] I have read the [Contributing Guidelines](https://github.com/HearthSim/Hearthstone-Deck-Tracker/blob/master/CONTRIBUTING.md#contributing)
- [ ] I have already signed the CLA

### CODE REVIEW

File: Hearthstone Deck Tracker/Hearthstone/Card.cs [hunk 1]
```diff
@@ -245,6 +246,8 @@ private string GetAlternativeText(bool formatted)
 
 		public int LettuceCooldown => _dbCard?.Entity.GetTag(GameTag.LETTUCE_COOLDOWN_CONFIG) ?? 0;
 
+		public int MultipleClasses => _dbCard?.Entity.GetTag(GameTag.MULTIPLE_CLASSES) ?? 0;
```
Comment (azeier) [inline Hearthstone Deck Tracker/Hearthstone/Card.cs:249]: This appears to be unused?

File: Hearthstone Deck Tracker/GameEventHandler.cs [hunk 2]
```diff
@@ -366,27 +366,36 @@ public void HandleTurnsInPlayChange(Entity entity, int turn)
...
 				return;
-			_game.Player.Class = hero!;
+			_game.Player.Class = heroName;
 			if(_game.CurrentGameStats != null)
-				_game.CurrentGameStats.PlayerHero = hero!;
-			Log.Info("Player=" + hero);
+			{
+				_game.CurrentGameStats.PlayerHero = heroName;
+				_game.CurrentGameStats.PlayerHeroCardId = cardId;
+				var hero= Database.GetCardFromId(cardId);
```
Comment (azeier) [inline Hearthstone Deck Tracker/GameEventHandler.cs:394]: Missing whitespace before the `=`

File: Hearthstone Deck Tracker/GameEventHandler.cs [hunk 3]
```diff
@@ -366,27 +366,36 @@ public void HandleTurnsInPlayChange(Entity entity, int turn)
...
-			Log.Info("Opponent=" + hero);
+			Log.Info("Opponent=" + heroName);
 		}
 
 		public void SetPlayerHero(string? cardId)
 		{
-			var hero = Database.GetHeroNameFromId(cardId);
-			if(string.IsNullOrEmpty(hero))
+			var heroName = Database.GetHeroNameFromId(cardId);
+			if(heroName == null)
```
Comment (azeier) [inline Hearthstone Deck Tracker/GameEventHandler.cs:387]: Is these a reason for this to not check `string.IsNullOrEmpty`?

File: Hearthstone Deck Tracker/GameEventHandler.cs [hunk 4]
```diff
@@ -366,27 +366,36 @@ public void HandleTurnsInPlayChange(Entity entity, int turn)
 
 		public void SetOpponentHero(string? cardId)
 		{
-			var hero = Database.GetHeroNameFromId(cardId);
-			if(string.IsNullOrEmpty(hero))
+			var heroName = Database.GetHeroNameFromId(cardId);
+			if(heroName == null)
```
Comment (azeier) [inline Hearthstone Deck Tracker/GameEventHandler.cs:370]: Is these a reason for this to not check `string.IsNullOrEmpty`?

**Review by azeier — CHANGES_REQUESTED** [review]
Did you test what the interaction with the "default" decks is? When you play a game without an active deck selected it goes into the "DefaultDeckStats" system, which is essentially an unnamed deck for each class.

 My matchup table currently looks like this
image

With these two matches in them
image

```xml
    <DeckStats>
      <DeckId>00000000-0000-0000-0000-000000000000</DeckId>
      <Games>
        <Game>
          <GameId>3150d0fa-d687-4a10-ba97-0f75cdfbd935</GameId>
          <PlayerHero>Diablo</PlayerHero>
          <OpponentHero>Rattlegore</OpponentHero>
          <Coin>true</Coin>
          <GameMode>Duels</GameMode>
          <Result>Loss</Result>
          <Turns>0</Turns>
          <StartTime>2021-11-17T16:18:28.1163677-08:00</StartTime>
          <EndTime>2021-11-17T16:18:38.019061-08:00</EndTime>
          <PlayerName>Epix#2966</PlayerName>
          <OpponentName>Rattlegore</OpponentName>
          <ReplayFile>Epix#2966(Diablo) vs Rattlegore(Rattlegore) 1618-171121.hdtreplay</ReplayFile>
          <WasConceded>true</WasConceded>
          <HearthstoneBuild>60915</HearthstoneBuild>
          <OpponentCardbackId>250</OpponentCardbackId>
          <FriendlyPlayerId>2</FriendlyPlayerId>
          <OpponentPlayerId>1</OpponentPlayerId>
          <ScenarioId>4429</ScenarioId>
          <ServerInfo>
            <Address>37.244.49.185</Address>
            <AuroraPassword>DPflFK</AuroraPassword>
            <ClientHandle>192564</ClientHandle>
            <GameHandle>16108</GameHandle>
            <Mission>4429</Mission>
            <Port>3724</Port>
            <Resumable>true</Resumable>
            <SpectatorMode>false</SpectatorMode>
            <SpectatorPassword>FIBKHn</SpectatorPassword>
            <Version>2540310</Version>
          </ServerInfo>
          <GameType>GT_PVPDR</GameType>
          <RankedSeasonId>97</RankedSeasonId>
          <OpponentHeroCardId>PVPDR_Hero_Rattlegore</OpponentHeroCardId>
          <Region>EU</Region>
          <DeckId>00000000-0000-0000-0000-000000000000</DeckId>
          <DeckName>none</DeckName>
          <HsReplay>
            <UploadId>XCsX4hLVNMX3NBQg6P486E</UploadId>
            <UploadTries>1</UploadTries>
            <Unsupported>false</Unsupported>
            <ReplayUrl>https://hsreplay.net/uploads/upload/XCsX4hLVNMX3NBQg6P486E/</ReplayUrl>
          </HsReplay>
          <PlayerCards>
            <Card Id="EX1_316" Count="1" />
            <Card Id="UNG_829" Count="1" />
            <Card Id="CORE_OG_241" Count="1" />
            <Card Id="SCH_702" Count="1" />
          </PlayerCards>
        </Game>
      </Games>
      <Name>Diablo</Name>
    </DeckStats>
```
```xml
    <DeckStats>
      <DeckId>00000000-0000-0000-0000-000000000000</DeckId>
      <Games>
        <Game>
          <GameId>40f3317b-1663-46a1-847e-e94189ac90c7</GameId>
          <PlayerHero>Turalyon, the Tenured</PlayerHero>
          <OpponentHero>Mozaki, Master Duelist</OpponentHero>
          <Coin>true</Coin>
          <GameMode>Duels</GameMode>
          <Result>Loss</Result>
          <Turns>2</Turns>
          <StartTime>2021-11-17T17:35:43.7579059-08:00</StartTime>
          <EndTime>2021-11-17T17:37:05.130384-08:00</EndTime>
          <PlayerName>Epix#2966</PlayerName>
          <OpponentName>Mozaki, Master Duelist</OpponentName>
          <ReplayFile>Epix#2966(Turalyon, the Tenured) vs Mozaki, Master Duelist(Mozaki, Master Duelist) 1737-171121.hdtreplay</ReplayFile>
          <WasConceded>true</WasConceded>
          <HearthstoneBuild>60915</HearthstoneBuild>
          <PlayerCardbackId>139</PlayerCardbackId>
          <OpponentCardbackId>157</OpponentCardbackId>
          <FriendlyPlayerId>1</FriendlyPlayerId>
          <OpponentPlayerId>2</OpponentPlayerId>
          <ScenarioId>4429</ScenarioId>
          <ServerInfo>
            <Address>37.244.49.206</Address>
            <AuroraPassword>DMoNeI</AuroraPassword>
            <ClientHandle>192564</ClientHandle>
            <GameHandle>16079</GameHandle>
            <Mission>4429</Mission>
            <Port>3724</Port>
            <Resumable>true</Resumable>
            <SpectatorMode>false</SpectatorMode>
            <SpectatorPassword>AonEIH</SpectatorPassword>
            <Version>2540310</Version>
          </ServerInfo>
          <GameType>GT_PVPDR</GameType>
          <RankedSeasonId>97</RankedSeasonId>
          <OpponentHeroCardId>PVPDR_Hero_Mozaki</OpponentHeroCardId>
          <Region>EU</Region>
          <DeckId>0beea5e0-56df-492f-9af3-c073e9be6518</DeckId>
          <DeckName>Duels Run 17-11 17:35</DeckName>
          <PlayerDeckVersion>
            <Build>0</Build>
            <Major>1</Major>
            <Minor>0</Minor>
            <Revision>0</Revision>
          </PlayerDeckVersion>
          <HsReplay>
            <UploadId>DzUM63GJqAM8DgPa34coxE</UploadId>
            <UploadTries>1</UploadTries>
            <Unsupported>false</Unsupported>
            <ReplayUrl>https://hsreplay.net/uploads/upload/DzUM63GJqAM8DgPa34coxE/</ReplayUrl>
          </HsReplay>
          <PlayerCards>
            <Card Id="CS2_087" Count="1" />
            <Card Id="OG_223" Count="1" />
            <Card Id="BAR_550" Count="1" />
            <Card Id="EX1_132" Count="1" />
            <Card Id="CS2_091" Count="1" />
            <Card Id="SCH_247" Count="1" />
            <Card Id="EX1_371" Count="1" />
            <Card Id="CORE_FP1_020" Count="1" />
            <Card Id="OG_198" Count="1" Unconfirmed="1" />
            <Card Id="UNG_961" Count="1" Unconfirmed="1" />
            <Card Id="SW_049" Count="1" Unconfirmed="1" />
            <Card Id="EX1_363" Count="1" Unconfirmed="1" />
            <Card Id="EX1_360" Count="1" Unconfirmed="1" />
            <Card Id="BAR_873" Count="1" Unconfirmed="1" />
            <Card Id="UNG_960" Count="1" Unconfirmed="1" />
            <Card Id="PVPDR_YOP_PaladinT1" Count="1" Unconfirmed="1" />
          </PlayerCards>
          <OpponentCards>
            <Card Id="SW_450" Count="1" />
            <Card Id="BAR_546" Count="1" />
          </OpponentCards>
        </Game>      
      </Games>
      <Name>Paladin</Name>
    </DeckStats>
```

File: Hearthstone Deck Tracker/Stats/GameStats.cs [hunk 5]
```diff
@@ -288,7 +288,11 @@ public int OpponentLegendRank
 
 		public int BrawlLosses { get; set; }
 
+		public string? PlayerHeroCardId { get; set; }
+		public string[]? PlayerHeroClasses { get; set; }
+
 		public string? OpponentHeroCardId { get; set; }
+		public string[]? OpponentHeroClasses { get; set; }
```
Comment (azeier) [inline Hearthstone Deck Tracker/Stats/GameStats.cs:295]: Add `ShouldSerialize` entries for these so we don't write a bunch of nulls to the game stats files. See https://github.com/HearthSim/Hearthstone-Deck-Tracker/blob/47d565774f7b03bf0482fa9a71b9d09695a96189/Hearthstone%20Deck%20Tracker/Stats/GameStats.cs#L563

File: Hearthstone Deck Tracker/Enums/HeroClass.cs [hunk 6]
```diff
@@ -52,6 +52,32 @@ public enum HeroClassAll
...
+		Priest,
+		[LocDescription("Enum_HeroClass_Rogue")]
+		Rogue,
+		[LocDescription("Enum_HeroClass_Shaman")]
+		Shaman,
+		[LocDescription("Enum_HeroClass_Warlock")]
+		Warlock,
+		[LocDescription("Enum_HeroClass_Warrior")]
+		Warrior,
+		[LocDescription("Enum_HeroClass_Neutral")]
```
[the remark this answers is not in the document]
Comment (edipo2s) [reply Hearthstone Deck Tracker/Enums/HeroClass.cs:77]: ah right, forgot to commit it

**Comment by edipo2s** [comment]
> Did you test what the interaction with the "default" decks is? When you play a game without an active deck selected it goes into the "DefaultDeckStats" system, which is essentially an unnamed deck for each class.
> 
> My matchup table currently looks like this image
> 
> With these two matches in them image
> 
> ```
>     <DeckStats>
>       <DeckId>00000000-0000-0000-0000-000000000000</DeckId>
>       <Games>
>         <Game>
>           <GameId>3150d0fa-d687-4a10-ba97-0f75cdfbd935</GameId>
>           <PlayerHero>Diablo</PlayerHero>
>           <OpponentHero>Rattlegore</OpponentHero>
>           <Coin>true</Coin>
>           <GameMode>Duels</GameMode>
>           <Result>Loss</Result>
>           <Turns>0</Turns>
>           <StartTime>2021-11-17T16:18:28.1163677-08:00</StartTime>
>           <EndTime>2021-11-17T16:18:38.019061-08:00</EndTime>
>           <PlayerName>Epix#2966</PlayerName>
>           <OpponentName>Rattlegore</OpponentName>
>           <ReplayFile>Epix#2966(Diablo) vs Rattlegore(Rattlegore) 1618-171121.hdtreplay</ReplayFile>
>           <WasConceded>true</WasConceded>
>           <HearthstoneBuild>60915</HearthstoneBuild>
>           <OpponentCardbackId>250</OpponentCardbackId>
>           <FriendlyPlayerId>2</FriendlyPlayerId>
>           <OpponentPlayerId>1</OpponentPlayerId>
>           <ScenarioId>4429</ScenarioId>
>           <ServerInfo>
>             <Address>37.244.49.185</Address>
>             <AuroraPassword>DPflFK</AuroraPassword>
>             <ClientHandle>192564</ClientHandle>
>             <GameHandle>16108</GameHandle>
>             <Mission>4429</Mission>
>             <Port>3724</Port>
>             <Resumable>true</Resumable>
>             <SpectatorMode>false</SpectatorMode>
>             <SpectatorPassword>FIBKHn</SpectatorPassword>
>             <Version>2540310</Version>
>           </ServerInfo>
>           <GameType>GT_PVPDR</GameType>
>           <RankedSeasonId>97</RankedSeasonId>
>           <OpponentHeroCardId>PVPDR_Hero_Rattlegore</OpponentHeroCardId>
>           <Region>EU</Region>
>           <DeckId>00000000-0000-0000-0000-000000000000</DeckId>
>           <DeckName>none</DeckName>
>           <HsReplay>
>             <UploadId>XCsX4hLVNMX3NBQg6P486E</UploadId>
>             <UploadTries>1</UploadTries>
>             <Unsupported>false</Unsupported>
>             <ReplayUrl>https://hsreplay.net/uploads/upload/XCsX4hLVNMX3NBQg6P486E/</ReplayUrl>
>           </HsReplay>
>           <PlayerCards>
>             <Card Id="EX1_316" Count="1" />
>             <Card Id="UNG_829" Count="1" />
>             <Card Id="CORE_OG_241" Count="1" />
>             <Card Id="SCH_702" Count="1" />
>           </PlayerCards>
>         </Game>
>       </Games>
>       <Name>Diablo</Name>
>     </DeckStats>
> ```
> 
> ```
>     <DeckStats>
>       <DeckId>00000000-0000-0000-0000-000000000000</DeckId>
>       <Games>
>         <Game>
>           <GameId>40f3317b-1663-46a1-847e-e94189ac90c7</GameId>
>           <PlayerHero>Turalyon, the Tenured</PlayerHero>
>           <OpponentHero>Mozaki, Master Duelist</OpponentHero>
>           <Coin>true</Coin>
>           <GameMode>Duels</GameMode>
>           <Result>Loss</Result>
>           <Turns>2</Turns>
>           <StartTime>2021-11-17T17:35:43.7579059-08:00</StartTime>
>           <EndTime>2021-11-17T17:37:05.130384-08:00</EndTime>
>           <PlayerName>Epix#2966</PlayerName>
>           <OpponentName>Mozaki, Master Duelist</OpponentName>
>           <ReplayFile>Epix#2966(Turalyon, the Tenured) vs Mozaki, Master Duelist(Mozaki, Master Duelist) 1737-171121.hdtreplay</ReplayFile>
>           <WasConceded>true</WasConceded>
>           <HearthstoneBuild>60915</HearthstoneBuild>
>           <PlayerCardbackId>139</PlayerCardbackId>
>           <OpponentCardbackId>157</OpponentCardbackId>
>           <FriendlyPlayerId>1</FriendlyPlayerId>
>           <OpponentPlayerId>2</OpponentPlayerId>
>           <ScenarioId>4429</ScenarioId>
>           <ServerInfo>
>             <Address>37.244.49.206</Address>
>             <AuroraPassword>DMoNeI</AuroraPassword>
>             <ClientHandle>192564</ClientHandle>
>             <GameHandle>16079</GameHandle>
>             <Mission>4429</Mission>
>             <Port>3724</Port>
>             <Resumable>true</Resumable>
>             <SpectatorMode>false</SpectatorMode>
>             <SpectatorPassword>AonEIH</SpectatorPassword>
>             <Version>2540310</Version>
>           </ServerInfo>
>           <GameType>GT_PVPDR</GameType>
>           <RankedSeasonId>97</RankedSeasonId>
>           <OpponentHeroCardId>PVPDR_Hero_Mozaki</OpponentHeroCardId>
>           <Region>EU</Region>
>           <DeckId>0beea5e0-56df-492f-9af3-c073e9be6518</DeckId>
>           <DeckName>Duels Run 17-11 17:35</DeckName>
>           <PlayerDeckVersion>
>             <Build>0</Build>
>             <Major>1</Major>
>             <Minor>0</Minor>
>             <Revision>0</Revision>
>           </PlayerDeckVersion>
>           <HsReplay>
>             <UploadId>DzUM63GJqAM8DgPa34coxE</UploadId>
>             <UploadTries>1</UploadTries>
>             <Unsupported>false</Unsupported>
>             <ReplayUrl>https://hsreplay.net/uploads/upload/DzUM63GJqAM8DgPa34coxE/</ReplayUrl>
>           </HsReplay>
>           <PlayerCards>
>             <Card Id="CS2_087" Count="1" />
>             <Card Id="OG_223" Count="1" />
>             <Card Id="BAR_550" Count="1" />
>             <Card Id="EX1_132" Count="1" />
>             <Card Id="CS2_091" Count="1" />
>             <Card Id="SCH_247" Count="1" />
>             <Card Id="EX1_371" Count="1" />
>             <Card Id="CORE_FP1_020" Count="1" />
>             <Card Id="OG_198" Count="1" Unconfirmed="1" />
>             <Card Id="UNG_961" Count="1" Unconfirmed="1" />
>             <Card Id="SW_049" Count="1" Unconfirmed="1" />
>             <Card Id="EX1_363" Count="1" Unconfirmed="1" />
>             <Card Id="EX1_360" Count="1" Unconfirmed="1" />
>             <Card Id="BAR_873" Count="1" Unconfirmed="1" />
>             <Card Id="UNG_960" Count="1" Unconfirmed="1" />
>             <Card Id="PVPDR_YOP_PaladinT1" Count="1" Unconfirmed="1" />
>           </PlayerCards>
>           <OpponentCards>
>             <Card Id="SW_450" Count="1" />
>             <Card Id="BAR_546" Count="1" />
>           </OpponentCards>
>         </Game>      
>       </Games>
>       <Name>Paladin</Name>
>     </DeckStats>
> ```

Stats for those default decks are working. What happens in your table was that you had a game with an old hero (Diablo) that was not in the hero classes mapping. I looked for old removed heroes but did not find any other, so I just add the Diablo that was missing there.

**Comment by anonymous555** [comment]
Does this support duels deck import too?

**Comment by edipo2s** [comment]
> Does this support duels deck import too?

Sorry, no. This is about add support for the Duels in the constructed matchups table

**Pull request merged**